fill in VolumeBrick for OSPRay to display read in data

demarle opened the pull request.

demarle committed:
fill in VolumeBrick for OSPRay to display read in data

mirenradia commented on modules/mpi/tutorials/ospMPIDistribTutorialVolumeHDF5.cpp:
```diff
@@ -90,13 +90,61 @@ int main(int argc, char **argv)
+    vec3f max{-MAXFLOAT};
+    for (auto x : hdf5reader.m_blockBounds) {
+        std::cerr << x.lower << " " << x.upper << std::endl;
+        if (x.lower.x < min.x)
+            min.x = x.lower.x;
+        if (x.lower.y < min.y)
+            min.y = x.lower.y;
+        if (x.lower.z < min.z)
+            min.z = x.lower.z;
+        if (x.lower.x > max.x)
+            max.x = x.lower.x;
+        if (x.lower.y > max.y)
+            max.y = x.lower.y;
+        if (x.lower.z > max.z)
+            max.z = x.lower.z;
```
It's possible this is quite right since the `int` indices correspond to counting cells (e.g. the lower corner of the cell) and presumably the float indices should correspond to the corners of the whole box so you need to add the cell width to each component of `max`.

mirenradia replied:
Any chance a similar mistake like this in the original gravity spheres example was causing your artifacts at the block boundaries you observed?

mirenradia replied:
See comment below.

demarle replied:
Could very well be one of the issues in the GSexample, I'll take a look shortly.

mirenradia commented on modules/mpi/tutorials/ospMPIDistribTutorialVolumeHDF5.cpp:
```diff
@@ -90,13 +90,61 @@ int main(int argc, char **argv)
+            max.y = x.lower.y;
+        if (x.lower.z > max.z)
+            max.z = x.lower.z;
+        myregions.push_back(box3f(x.lower, x.upper));
```
I think this is wrong as all ranks store all block bounds. This should probably be conditioned on `ChomboHDF5::Reader::m_rankDataOwner == rank` or something.

mirenradia replied:
I have added a function: `ChomboHDF5::Reader::getMyRegions` which should return the object you need for myregions (i.e. only the bounds of the boxes on this rank).

demarle replied:
updated the viewer commit to use getMyRegions. ty

mirenradia commented on modules/mpi/tutorials/ospMPIDistribTutorialVolumeHDF5.cpp:
```diff
@@ -90,13 +90,61 @@ int main(int argc, char **argv)
+    brick.brick = hdf5reader.getVolume();
+    brick.bounds = myregions;
+
+    worldBounds = box3f(min*0.25, max*0.25); //seems like we are missing the block scale somewhere
```
For some reason I thought this could be easily deduced from what is read in `ChomboHDF5::Reader` but now the only way I can think of doing so is from iterating through all the boxes on the coarsest level and finding the one with the largest upper corner which is a bit of a faff. This information is stored in the HDF5 file so I will add some code to read it in.

mirenradia replied:
Ok, I've now added a function `ChomboHDF5::Reader::getDomainBounds` which should return a suitable `box3f` you can set the `worldBounds` to.

demarle replied:
updated, ty.

mirenradia commented on modules/mpi/tutorials/ospMPIDistribTutorialVolumeHDF5.cpp:
```diff
@@ -90,13 +90,61 @@ int main(int argc, char **argv)
+    std::vector<float> opacities = {0.0f, 0.5f, 1.0f};
+    tfn.setParam("color", cpp::CopiedData(colors));
+    tfn.setParam("opacity", cpp::CopiedData(opacities));
+    vec2f valueRange = vec2f(0, 4080.02);
```
Is it worth adding a feature in the reader to determine the extremal values for the read in data?

demarle replied:
yes, because of the opaque API, it is harder than I'd want to query the ranges after we make the underlying array. So yes please do add it to the reader. Make it optional if you like as an optimization path when we happen to know the range already.

mirenradia replied:
Done. It's on my branch as `ChomboHDF5::Reader::calculateValueRange()`.

mirenradia reviewed (commented):
Hi @demarle, I found a few potential issues and few things for me to add to the reader. Have a look at the comments below and let me know what you think.

demarle committed:
incorporate reader improvements

demarle committed:
WIP: call wombat to convexify the AMR to make compositing possible

Exporting more (all?) of the metadata so that wombat can
know where all the blocks are and who owns them etc.

Image is showing progress, add the -noWombatNoCry arg to disable
wombat and see what it looks like with ray module_mpi.

mirenradia commented:
I'm fine with the modifications to the reader and can't see anything obviously wrong.

demarle committed:
closer to working

correct the calculation of block boundaries, and thus regions, that
we get out of wombat. With this single rank image is now correct and
multirank runs now have things in right place but show artifacts
between blocks

mess with transfer function a bit more to make it close

update the samplerate parameter destination and name for osp2x.
(commented out)

restored original distributed volume tutorial but fixed the missing
gaps between ranks by being sure to allocate the outermost edge of voxels.
(see nVoxels change).

demarle committed:
fix per level spacing mostly

fixes placement of blocks in scene, for real this time, thus no more small ghosts.
added a half voxel nudge to clipping that makes amelia's data almost
perfect.

remove code leftover from original file

messing with debugfs

demarle committed:
deworm wombat some more

the parent/child thing wasn't right on this data at least,
so the output wasn't actually convex. This fixes that.

there are many many blocks now, so render performance suffers

demarle committed:
the +1 left over from first draft of geneology was wrong

it was wrong and lead to an order of magnitude more unit bricks
than needed

demarle committed:
use NaN for non local data, and try to tweek appearance

demarle committed:
improve the name of the Chombo example

demarle committed:
start on an even simple example for debugging with

and remove some dead code

demarle committed:
finish the small simple example

seems to be more or less correct. An observation is that
remote value treatment is tricky, NAN is the worst, the original
approach of setting remotes to 0 and letting the LUT handle it
is pretty good but boundaries are fairly visible where they
transition to the 0 color. Setting remote values to local value is
quite close, but there are still visible boundaries. Setting it to
the actual value is closer still but even that has some artifacts.
